content/blog/2021-12-21-December-21-community-gems.md

```diff
@@ -0,0 +1,174 @@
+
+That's a really good question @Kwon-Young!
+
+You can always look through experiment metrics with `dvc exp show` and this
+shows you all of the experiments you've run.
+
+To get the metrics for a specific experiment or set of experiments, you'll need
+the experiment ids and then you can use the Python API like this example.
+
+```python
+from dvc.repo import Repo
+
+dvc = Repo(".")  # or Repo("path/to/repo/dir")
+metrics = dvc.metrics.show(revs=["exp-name1", "exp-name2", ...])
+```
+
+This returns a Python dictionary that contains what gets displayed in
+`dvc metrics show --json` except you're able to specify the experiments you
+want.
+
+### [If I have a DVC pipeline with many stages as well as `foreach` in every stage, is it possibile to run the whole pipeline but only for one element of the `foreach`?](https://discord.com/channels/485586884165107732/563406153334128681/915986804577026088)
+
+Another great question from @vgodie!
+
+If your stages look something like this for example:
+
+```yaml
+stages:
+  cleanups:
+    foreach: # List of simple values
+      - raw1
+      - labels1
+      - raw2
+    do:
+      cmd: clean.py "${item}"
+      outs:
+        - ${item}.cln
+  train:
+    foreach:
+      - epochs: 3
+        thresh: 10
+      - epochs: 10
+        thresh: 15
+    do:
+      cmd: python train.py ${item.epochs} ${item.thresh}
+```
+
+You should try the following command:
+
+```dvc
+$ dvc repro cleanups@labels1
+```
+
+This will run your whole pipeline, but only with `labels1` in the `cleanups`
+stage.
+
+### [Is it possible to pull experiments from the remote without checking out the base commit of those experiments?](https://discord.com/channels/485586884165107732/485596304961962003/910481311905505290)
+
+Thanks for the question @mattlbeck!
+
+You should be able to do this with `dvc exp pull origin exp-name`.
+
+If you have experiments with the same name on different commits, using
+`exp-name` won't work since it defaults to selecting the one based on your
+current commit if there are duplicate names.
+
+To work around this, you can use the full refname, like
+`refs/exps/e7/78ad744e8d0cd59ddqc65d5d698cf102533f85/exp-6cb7`, to specify the
+experiments that you want to work with.
+
+### [How should I handle checkpoints in PyTorch Lightning with DVCLive?](YouTube link here)
+
+This is a really good question that came from one of our Office Hours talks!
+Thanks Ilya Sirotkin!
+
+We have an [open issue](https://github.com/iterative/dvclive/issues/170) we
+encourage you to follow for more details and to even contribute!
+
+Python Lightning handles callbacks differently from many other libraries. This
+affects the way metrics logging is executed and how models are saved.
+
+You can write a custom callback to control saving everything and track it with
+DVC and this is the workaround we suggest. You can implement the
+`after_save_checkpoint` method and save the model file.
```

Comment:
Does this fully represent his workaround @Davoo?  Do we want to show the pic with the ins outs?

Reply:
Yes, please let me know if that is enough for the workaround. I wasn't quite sure what else to add there. 😅

Reply:
Not fully. Ilia's suggested adapting your code and pipeline to make use of DVC `foreach` and reuse the `ModelCheckpoint` included in `pytorch_lightning`.

Full description of Ilia's suggestion:

<details>

> break your training process into small stages, which you will not be afraid to lose. You should specify the stage’s checkpoint as the output of the stage and set it as a dependency for the next stage. Thus, if something breaks, the dvc repro command will resume your experiment from the last stage.


```yaml
stages:
  stage_0:
    cmd: python train.py
    outs:
      - checkpoints/checkpoint_epoch=0.ckpt
  next:
    foreach:
      - prev: 0
        next: 1
      - prev: 1
        next: 2
    do:
      cmd: python train.py --checkpoint ${item.prev}
      deps:
        - checkpoints/checkpoint_epoch=${item.prev}.ckpt
      outs:
        - checkpoints/checkpoint_epoch=${item.next}.ckpt
```

```python
import os
import pytorch_lightning as pl
import argparse
from dvclive.lightning import DvcLiveLogger
from tests.test_lightning import LitMNIST

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--checkpoint', dest='prev_epoch', action='store', default=-1,
                        help='Previous epoch number')
    parser.add_argument('--epochs', dest='run_epochs', action='store', default=1,
                        help='Number of epochs to run in this iteration')
    args = parser.parse_args()

    model = LitMNIST()
    # set checkpoint path
    ckpt_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "checkpoints"))
    # checkpoints will be saved to checkpoints/checkpoint_epoch={epoch_number}.ckpt
    cp = pl.callbacks.model_checkpoint.ModelCheckpoint(
        monitor="train_loss_epoch",
        save_top_k=1,
        dirpath=ckpt_path,
        filename='checkpoint_{epoch}')

    # init logger
    dvclive_logger = DvcLiveLogger()
    # set None if it is the first run
    resume_from_checkpoint = None if int(args.prev_epoch) < 0 else os.path.join(ckpt_path, f"checkpoint_epoch={args.prev_epoch}.ckpt")
    # increase maximum number of epochs to run current iteration
    max_ep = max(1, int(args.prev_epoch) + 1 + int(args.run_epochs))
    trainer = pl.Trainer(
        logger=dvclive_logger, max_epochs=max_ep,
        callbacks=[cp],
        resume_from_checkpoint=resume_from_checkpoint
    )
    trainer.fit(model)
```
</details>

If this is too long to include, I can add this suggestion as a comment in the `dvclive` issue so the community gem post could just link .

Reply:
I included pieces of the code, but I think it'll be better to have the entire workaround as a comment in the `dvclive` issue.

Reply:
@daavoo can it be done simpler with persistent outputs? why the common way dvc/dvclive handle checkpoints doesn't work? Could you please give a bit more context? Is there a ticket for this?

Reply:
@shcheklein Here's a link to this issue: https://github.com/iterative/dvclive/issues/170

This seems like a discussion better for that issue than here in community gems.